### PR TITLE
Media-ranker-Gessica Mohr

### PR DESCRIPTION
**Pull request by GeMath18**
# Media Ranker
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
---      | ---
Describe a custom model method you wrote. | 
Describe how you approached testing that model method. What edge cases did you come up with? | 
What are `session` and `flash`? What is the difference between them? | 
What was one thing that you gained more clarity on through this assignment? | 
What is the Heroku URL of your deployed application? | 

# Assignment Submission: Media Ranker

Congratulations! You're submitting your assignment. Please reflect on the assignment with these questions.

## Reflection

Prompt | Response
--- | ---
What was a custom model method you wrote? What was it responsible for doing? | I created the top_ten method that is responsible to show the top 10 works of each category on the same page. 
Describe how you approached testing that model method. What edge cases did you come up with? | i tested to make sure it would return up to 10 (<=10) for each category. Also tested to make sure it would return an empty array if there are works (movies,albums, books).
What are `session` and `flash`? What is the difference between them? | Flash is a type of hash that sends one time messages from the controller to the views. For example if people create a new book or movie, it populates a message that the action was created and it successfully the information was updated or shows an error message that it wasn't done. Session is also a type of hash that keeps track of data when the user has opened a session. It is used to store the users' information when they log in. The difference between them is that session stores the data and will not go away after the next request-response cycle. Instead it will be stored indefinitely. The data will be deleted when the user closes the browser or remove the data when the user logs out. 
What was one thing that you gained more clarity on through this assignment? | I definitely got more practice using MVC, how all comes together at the end. I got more clarity on how Active Record Relationships works when we create a model and how it relates to other model classes. in this cate vote , belongs to user and belongs to work. I generated two migrations that relateWorkstoVotes and relateUserstoVotes. this made the table Votes create a work_id and user_id to connect with them. Got more practice using html and CSS. I really like this project! 
What is the Heroku URL of your deployed application? | https://mr-gessica-app.herokuapp.com

### CODE REVIEW

File: app/controllers/works_controller.rb [hunk 1]
```diff
@@ -0,0 +1,78 @@
+class WorksController < ApplicationController
+  def index
+    @works = Work.all
+  end
+
+  def show
+    work_id = params[:id]
+    @work = Work.find_by(id: work_id)
+    if @work.nil?
+      head :not_found
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:10]: Consider using a `redirect` instead of `head :not_found` so you can redirect the user to somewhere useful :)

File: app/controllers/works_controller.rb [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+  end
+
+  def destroy
+    work_id = params[:id]
+    @work = Work.find_by(id: work_id)
+
+    if @work.nil?
+      head :not_found
+      return
+    end
```
Comment (beccaelenzil) [inline app/controllers/works_controller.rb:64]: Consider including this code in a controller filter to DRY up your code.

File: config/routes.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+Rails.application.routes.draw do
+  # For details on the DSL available within this file, see https://guides.rubyonrails.org/routing.html
+
+  root to: 'homepages#index'
+  resources :homepages, only: [:index]
+  resources :works do
+    resources :votes, only: [:create]
```
Comment (beccaelenzil) [inline config/routes.rb:7]: Nice use of a nested route for the upvote feature.

File: app/views/homepages/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        <p class="spotlight__description">
+          <%= Work.spotlight.votes.count %> votes - <%= Work.spotlight.description %>
+        </p>
+      <% end  %>
+      </section>
+
+  <hr class="root__hr">
+  <section class="top-ten__container">
+    <% hash = { "movie" => "Top Movies", "album" => "Top Albums", "book" => "Top Books" } %>
+    <% ["movie", "book", "album"].each do |category| %>
```
Comment (beccaelenzil) [inline app/views/homepages/index.html.erb:17]: Great use of a loop to DRY up your code.

File: app/views/works/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+        <tr>
+          <th width="80">Votes</th>
+          <th>Title</th>
+          <th width=28%>Created By</th>
+          <th width="100">Published</th>
+          <th width="100">Upvote</th>
+        </tr>
+        </thead>
+        <tbody>
+        <% works.each do |work| %>
```
Comment (beccaelenzil) [inline app/views/works/index.html.erb:22]: Again, great use of a loop to DRY up your code. Consider how you could also do this with a partial view.

File: app/models/work.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+
+  def self.spotlight
+    works = Work.all.max_by { |work| work.votes.count}
+  end
+
+  def self.top_ten(category)
+    works = Work.all.where(category: category)
+    return works.max_by(10) { |work| work.votes.count}
+  end
```
Comment (beccaelenzil) [inline app/models/work.rb:17]: Good work putting the business logic in the model and writing clear and concise methods.

**Review by beccaelenzil — COMMENTED** [review]
# Media Ranker





## Functional Requirements: Manual Testing

| Criteria | yes/no |
| --- | --- |
| **Before logging in** | --
| 1. On index page, there are at most 10 pieces of media on three lists, and a Media Spotlight | ✔️
| 2. Can go into a work's show page | ✔️
| 3. Verify unable to vote on a work, and get a flash message | ✔️
| 4. Can edit this work successfully, and get a flash message | ✔️
| 5. Can go to "View all media" page and see three lists of works, sorted by vote | ✔️
| 6. Verify unable to create a new work when the form is empty, and details about the validation errors are visible to the user through a flash message | ✔️
| 7. Can create a new work successfully. Note the URL for this work's show page | ✔️
| 8. Can delete this work successfully |✔️
| 9. Going back to the URL of this deleted work's show page produces a 404 or some redirect behavior (and does not try to produce a broken view) | ✔️
| 10. Verify that the "View all users" page lists no users | ✔️
| **Log in** | --
| 11. Logging in with a valid name changes the UI to "Logged in as" and "Logout" buttons | ✔️
| 12. Your username is listed in "View all users" page | ✔️
| 13. Verify that number of votes determines the Media Spotlight | ✔️
| 14. Voting on several different pieces of media affects the "Votes" tables shown in the work's show page and the user's show page | ✔️
| 15. Voting on the same work twice produces an error and flash message, and there is no extra vote | ✔️
| **Log out** | --
| 16. Logging out showed a flash message and changed the UI | ✔️
| 17. Logging in as a new user creates a new user | ✔️
| 18. Logging in as an already existing user has a specific flash message | ✔️

## Major Learning Goals/Code Review

| Criteria | yes/no |
| --- | --- |
| 1. Sees the full development cycle including deployment, and the app is deployed to Heroku | ✔️
| 2. Practices full-stack development and fulfilling story requirements: the styling, look, and feel of the app is similar to the original Media Ranker | ✔️
| 3. Practices git with at least 25 small commits and meaningful commit messages | ✔️

### Previous Rails learning, Building Complex Model Logic, DRYing up Rails Code

| Criteria | yes/no |
| --- | --- |
| 4. Routes file uses `resources` for works | ✔️
| 5. Routes file shows intention in limiting routes for voting, log-in functionality, and users | ✔️
| 6. The homepage view, all media view, and new works view use semantic HTML | ✔️
| 7. The homepage view, all media view, and new works view use partials when appropriate | ✔️
| 8. The model for media (likely named `work.rb`) `has_many` votes | ✔️
| 9. The model for media has methods to describe business logic, specifically for top ten and top media, possibly also for getting works by some category | ✔️
| 10. Some controller, likely the `ApplicationController`, has a controller filter for finding a logged in user | ✔️
| 11. Some controller, likely the `WorksController`, has a controller filter for finding a work | See in-line comment.
| 12. The `WorksController` uses strong params | ✔️
| 13. The `WorksController`'s code style is clean, and focused on working with requests, responses, `params`, `session`, `flash` | ✔️

### Testing Rails Apps

| Criteria | yes/no |
| --- | --- |
| 14. There are valid fixtures files used for users, votes, and works | ✔️
| 15. User model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 16. Vote model has tests with sections on validations (valid and invalid) and relationships (belongs to a user, belongs to a vote) | ✔️
| 17. Work model has tests with sections on validations (valid and invalid) and relationships (has votes) | ✔️
| 18. Work model has tests with a section on all business logic methods in the model, including their edge cases | ✔️

## Overall Feedback

Great job overall! Your implementation matches the demo site very closely, and the learning goals for this assignment were definitely met. Your test fixtures are very clean and allow for thorough testing of the validations and relationships. Your code is clear, readable, and well tested. I've left a few inline comments for you to review. Please let me know if you have any questions. Keep up the hard work!

| Overall Feedback | Criteria | yes/no |
| --- | --- | --- |
| Green (Meets/Exceeds Standards) | 14+ in Functional Requirements: Manual Testing && 14+ in Code Review | ✔️
| Yellow (Approaches Standards) | 12+ in Functional Requirements: Manual Testing && 11+ in Code Review, or the instructor judges that this project needs special attention | 
| Red (Not at Standard) | 0-10 in Code Review or 0-11 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention | 





## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality | Yes? |
| --- | --- |
| Perfect Indentation | ✅
| Elegant/Clever | ✅
| Descriptive/Readable | ✅
| Concise | ✅
| Logical/Organized | ✅
